### PR TITLE
database functions retrieve matches dynamically

### PR DESCRIPTION
**Pull request by yvesarrata**
hello.
A few functions were added to the database to ease querying matches.
Matches are retrieved from the db to be shown in the match list of the active user.

Something I noticed is that matches are retrieved each time the user goes back to his list of matches. I think that comes from conditional rendering of our components. Since they're always re rendered,  "onMount" is called multiple times hence db overloading...
Do you know if there's a way to use react-router to render each component once and then just redirect the user according to where he clicks on the app ?

Bises

### CODE REVIEW

**Comment by yvesarrata** [comment]
solved the issue mentioned above by querying db in the app component instead.

File: src/App.tsx [hunk 1]
```diff
@@ -5,10 +5,24 @@ import Header from './Header';
 import Login from './Login';
 import React, { SetStateAction, useEffect, useState } from 'react';
 import { supabase } from './setupSupabase';
+import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
 
 function App() {
   const { mode, renderHeader } = Header();
   const [session, setSession] = useState(null);
+  const [userId, setUserId] = useState<string>('');
```
Comment (nyc4m) [inline src/App.tsx]: For all calls to the db it's better to encapsulate them inside hooks for 2 reasons : 
1. it allow for maximum reusability
2. It "hides" supabase from the components, so if we wanna ditch supabase later it's gonna be easier (we never know)

Other than that nice !

File: src/App.tsx [hunk 2]
```diff
@@ -5,10 +5,24 @@ import Header from './Header';
...
+  const [matches, setMatches] = useState<Array<any>>([]);
+
+  const getMatch = async () => {
+    try {
+      const { data, error } = await supabase.rpc('f_getmatchinfo', {
```
Comment (nyc4m) [inline src/App.tsx]: I think the creation of a postgres function to do a join is a bit overkill, what benefits do you see doing so ?

**Review by nyc4m — CHANGES_REQUESTED** [review]
It would be better to create some hooks for db access, but other than that it seems cool 🔥

**Comment by nyc4m** [comment]
@sylvainPanneau for network optimization we will see later :)

**Comment by yvesarrata** [comment]
> It would be better to create some hooks for db access, but other than that it seems cool fire

yeah will do good idea

> I think the creation of a postgres function to do a join is a bit overkill, what benefits do you see doing so ?

better re usability and the functions will evolve so it's easier to read imo

**Comment by yvesarrata** [comment]
@nyc4m how do you like it now ?

File: src/database/triggers.sql [hunk 3]
```diff
@@ -8,3 +8,22 @@ $func_create_user$ LANGUAGE plpgsql security definer;
 DROP trigger if exists trigger_create_user on auth.users;
 CREATE TRIGGER trigger_create_user after insert on auth.users
     FOR EACH ROW EXECUTE PROCEDURE func_create_user();
+
+
+CREATE OR REPLACE FUNCTION func_match_notification() RETURNS trigger AS $func_match_notification$
+    BEGIN
+        update public.users
+        set "matchNotification" = "matchNotification" + 1
```
Comment (yvesarrata) [inline src/database/triggers.sql]: I have added a column called matchNotification in public.users table.
This attribute is updated by a trigger each time an insert in match is performed.
The update affects the users concerned by the new match relationship.
Then, a listener (websocket) is installed in useDatabase.ts/useUserId/matchInsert to listen for updates on this counter.
That way, we can react to a new match for a specific user and update the match list accordingly.
This whole trick was implemented because of the lack of filtering for insert listeners with supabase.
Long story short : you can't listen for an insert event for one and only one user because, well, it is not in the db yet...
Listening for changes in the match table and then filtering for the logged in user would be performance heavy if lots of matches happen at the same time across our future 10k + user base 💯

**Comment by nyc4m** [comment]
> better re usability and the functions will evolve so it's easier to read imo

@sylvainPanneau problem I see with this approach is that a part of the code is going to be outside of the repo (because it's declared directly on supabase) and it's going to decrease readability. Otherthing is that the supabase api allows you to have more simple joins (which are going to automatically fetch relationship between tables) and this is not something you'll have using sql functions. 
I do think functions can come in handy, but IMHO not in your use case. for reusability encapsulate in hooks is a better option i think

**Comment by yvesarrata** [comment]
@nyc4m ton grand père n'hésite pas à me dire ce que t'en penses

**Review by nyc4m — APPROVED** [review]
